js/kraken.js

```diff
@@ -1063,6 +1064,28 @@ module.exports = class kraken extends Exchange {
         return this.extend ({ 'info': response }, order);
     }
 
+    async fetchOrderTrades (id, symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        const response = await this.privatePostQueryOrders (this.extend ({
+            'trades': true, // whether or not to include trades in output (optional, default false)
+            'txid': id, // do not comma separate a list of ids - use fetchOrdersByIds instead
+            // 'userref': 'optional', // restrict results to given user reference id (optional)
+        }, params));
+        const tradeIds = response['result'][id]['trades'];
+        let trades = {};
+        const batchSize = 20; // maximum trade query request size
+        for (let i = 0; i < tradeIds.length; i += batchSize) {
+            const tradeIdBatch = tradeIds.slice (i, i + batchSize);
+            const tradeResponse = await this.privatePostQueryTrades ({ 'txid': tradeIdBatch.join (',') });
```

Comment:
The problem is that doing more than one HTTP request per unified call breaks the rate-limiting and pagination. This is why we don't loop HTTP requests inside the unified methods Instead, we should rather remove the `privatePostQueryOrders` call from this method. And leave just the `privatePostQueryTrades` call, with a requirement for `txid` params override (see other exchanges throwing `ArgumentsRequired`: https://github.com/ccxt/ccxt/blob/master/js/bitmart.js#L761)

Reply:
Gotcha, thought that request loop might be problematic with consistency.

That said, the preposed solution obviously detours from the unified interface a bit since the `id` (orderId) is no longer relevant or required and all that would be required is a concatenated string for `params.txid`.

And assuming on is trying to get tradeIds from the order. They'd have to use `fetchOrder` then look inside the `info` property to find the raw `info.trades` array of Ids.

Wouldn't it be possible (maybe difficult?) to include any calls to `privatePost` / `publicGet` to be queued within the rate limiter?

That way more advanced calls could be done and we can often keep more in line with the unified design.

Thanks!

Reply:
If the problem is that the unified method calls are rate limited, and each of them use a `privatePost` / `publicGet` call so therefore would be doubly rate limited.

I'd suggest maybe the consistent solution would be that only the `privatePost` and `publicGet` calls should be rate limited and the unified calls not. (obviously they'll always `await` the native calls).

As the native calls really are the only calls that need rate limiting since they're the ones being sent to the exchanges.

Anyway... Just a thought, I don't have nearly the depth and understanding of the library as you do.

Reply:
> Wouldn't it be possible (maybe difficult?) to include any calls to privatePost / publicGet to be queued within the rate limiter?

Those calls are all going through the rate-limiter already, however, the problem is a bit deeper. In general, the user would expect a defined execution time from a call (timeout), and the loops break the pagination as well. Suppose that the user doesn't know anything about the inner workings of fetchOrderTrades in this case – the user might call it in parallel multiple times, that will effectively launch multiple concurrent loops, which is harder to manage if the user wants to be in precise control of his rate limits.

> > They'd have to use fetchOrder then look inside the info property to find the raw info.trades array of Ids.

I would suggest a different solution. First, we need to add the parseTrades logic to `parseOrder` – that way `fetchOrder` will return an order containing an array of unified `'trades'` (no need for `info` lookups), each trade having just the id (all other fields undefined). From there the second step is to add the support for fetchOrderTrades like so:

```
const order = await kraken.fetchOrder (id)  // ←------------------------------ ↓
const orderTrades = await kraken.fetchOrderTrades (order['id'], since, limit, order)
```

↑ Apart from that we should indicate, that Kraken does not have a true endpoint for order trades and may require multiple calls, therefore we are emulating it, so `has['fetchOrderTrades']` must be set to `'emulated'`, instead of `true`.

Also, the batch size has to be put inside options to keep pagination configurable.

> That way more advanced calls could be done and we can often keep more in line with the unified design.

We want the lib to be a plain utilitarian tool (as close to one http request per call as possible), so that it would offer building blocks for complex functions, but would not do the complex functions inside the lib. So we tend to keep it as low-level as possible, instead of making it a swiss-knife for everything that would barely work. We tend to not emulate methods that are not exposed by the exchange API the way they should be exposed. Otherwise we will encourage the exchanges to design nonsense APIs and will just drown in supporting all the quirks, because of the exchange specifics.

Reply:
@roark I'll fix the PR and will merge asap, thank for your help and for the feedback!

Reply:
@kroitor Thanks! Greatly appreciate the quick responses and excellent library!